include/deal.II/base/tensor.h

```diff
@@ -106,62 +348,80 @@ class Tensor
   typedef typename numbers::NumberTraits<Number>::real_type real_type;
 
   /**
+   * Type of objects encapsulated by this container and returned by
+   * operator[](). This is a tensor of lower rank for a general tensor, and
+   * a scalar number type for Tensor<1,dim,Number>.
+   */
+  typedef typename Tensor<rank_-1,dim,Number>::tensor_type value_type;
+
+  /**
    * Declare an array type which can be used to initialize an object of this
    * type statically.
    */
-  typedef typename Tensor<rank_-1,dim,Number>::array_type array_type[dim];
+  typedef typename Tensor<rank_-1,dim,Number>::array_type
+  array_type[(dim != 0) ? dim : 1];
+  // ... avoid a compiler warning in case of dim == 0 and ensure that the
+  // array always has positive size.
 
   /**
-   * Constructor. Initialize all entries to zero.
+   * Constructor. Initialize all entries to zero if
+   * <tt>initialize==true</tt>; this is the default behaviour.
    */
-  Tensor ();
+  explicit
+  Tensor (const bool initialize = true);
```

Comment:
I think the current state is actually quite good, because compilers can decompose these arrays into individual scalars and their initialization is explicit for each element (as opposed to more difficult to figure out if you had a memset or similar that initializes *all* elements at once). The point being that if you had explicit initialization of individual scalars, the compiler can easily forward propagate this and, for example, if you happened to immediately store something in the tensor (as one often does) throw away the original initialization as a dead store. That would be much harder if you had a memset.